Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,234 @@
+
+Events are divided into two categories: **known** events and **dynamic** (or **custom**) events. Known events are known
+to the EE and correspond one-to-one to individual event types fired by the underlying platform loggers. Dynamic events are
+events not known to the EE; their use and description is below in the "Dynamic Events" section.
+
+All events are fired through the `IGCToCLREventSink` interface, which is accessed through the `IGCToCLR` interface given
+to the GC on startup:
+
+```c++
+class IGCToCLREventSink
+{
+
+};
+
+class IGCToCLR
+{
+    virtual IGCToCLREventSink* EventSink() = 0;
+};
+```
+
+Every known event is fired through its own dedicated callback on `IGCToCLREventSink`. For example, the `GCEnd` event
+is fired through a callback like this:
+
+```c++
+class GCToCLREventSink : public IGCToCLREventSink { ... }
+
+GCToCLREventSink::FireGCEnd(uint32_t count, uint16_t depth)
+{
+    // ...
+}
+```
+
+`GCTOCLREventSink::FireGCEnd` is responsible for constructing and dispatching the event to platform loggers if the event
+is enabled. The principal advantage of having one callback per known event is that known events can reach into EE internals
+and add data to events that the GC otherwise would not be aware of. Concrete examples of this would be:
+* The addition of `ClrInstanceId` to the payload of many events fired by the GC
+* Getting human-readable type names for objects allocated on the heap
+* Correlating `GCStart` events with collections induced via ETW.
+
+## Defining and Firing Dynamic Events
+
+It is useful for a standalone GC to be able to fire events that the EE was not previously aware of. For example, it is useful for a GC developer to add some event-based instrumentation to the GC, especially when testing new features and ensuring that they work as expected. Furthermore, it is desirable for GCs that are shipped from this repository
+(the "CLR GC") to interopate seamlessly with future versions of the .NET Core EE, which implies that it should be possible
+for the GC within this repository to add new events without having to recompile the runtime.
+
+While it is possible for some eventing implementations to receive events that are created at runtime, not all eventing implementations (particularly LTTNG) are not flexible enough for this. In order to accomodate new events, another method is added to `IGCToCLREventSink`:
+
+```c++
+void IGCToCCLREventSink::FireDynamicEvent(
+    /* IN */ const char* eventName,
+    /* IN */ void* payload,
+    /* IN */ size_t payloadSize
+);
+```
+
+A runtime implementing this callback will implement it by having a "catch-all" GC event whose schema is an arbitrary sequence of bytes. Tools can parse the (deliberately unspecified) binary format provided by GC events that use this mechanism in order to recover the data within the payload.
+
+Of note with dynamic events is that the CLR GC *may* ship with them. It is highly desirable that the CLR GC be able 
```

Comment:
I find this paragraph quite confusing. these are the points I'd convey:

1) adding a dynamic event is normal thing to do; as we've basically shipped new ETW events with every version of CLR;
2) it's not that whether GC will "tolerate a recompile of the runtime" (I am not sure what that means really); the only times we'd ask that a user must get a newer version of clr.dll to go with clrgc.dll are when something really incompatible has happened that breaks things functionally, eg, object layout changed in a way that GC is now not reading the correct obj size or ref sites; or if you want a new GC feature that lives in clrgc.dll to work functionally you must get the correspondingly newer version of clr.dll. and adding new events is definitely not one of them. so I am not sure why we'd need to periodically convert dynamic events to known events. if tools are already interpreting these dynamic events, they can keep interpreting them.

Reply:
Will do! I figured that having large numbers of dynamic events is not a good thing - I personally figured that favoring schematized events over unschematized ones makes sense for events that we intend to keep around long-term, which would be the case for dynamic events that have been shipped a few times.

At some point we are going to need to make major changes to the GC/EE interface, which is what I would mean by "tolerating a recompile".

Reply:
I see. how do you envision tools would use this if we change from dynamic events to "known events". does this mean tooling will need to change (even for PerfView, do we expect to interpret both the "dynamic" and the "known" version of the same event)? or somehow they light up because they are now known? since this is for tooling, it would be useful to explain how tools can use it.

Reply:
The tooling would need to change, yes. If we intend expect to change dynamic events to known events at any point, we'd need to teach PerfView to be aware of the known version of the event.